core/analysis/ngram_token_stream.cpp

```diff
@@ -139,25 +169,21 @@ NS_ROOT
 NS_BEGIN(analysis)
 
 /*static*/ analyzer::ptr ngram_token_stream::make(
-    size_t min, size_t max, bool preserve_original
+    const options_t& options
 ) {
-  return std::make_shared<ngram_token_stream>(min, max, preserve_original);
+  return std::make_shared<ngram_token_stream>(options);
 }
 
 /*static*/ void ngram_token_stream::init() {
   REGISTER_ANALYZER_JSON(ngram_token_stream, make_json); // match registration above
 }
 
 ngram_token_stream::ngram_token_stream(
-    size_t min_gram,
-    size_t max_gram,
-    bool preserve_original
+    const options_t& options
 ) : analyzer(ngram_token_stream::type()),
-    min_gram_(min_gram),
-    max_gram_(max_gram),
-    preserve_original_(preserve_original) {
-  min_gram_ = std::max(min_gram_, size_t(1));
-  max_gram_ = std::max(max_gram_, min_gram_);
+    options_(options) {
+  options_.min_gram = std::max(options_.min_gram, size_t(1));
```

Comment:
wouldn't be better if options itself will be responsible for ensuring data correctness, e.g.
```
struct options_t {
  options_t(size_t min_gram, size_t max_gram)
    : min_gram(std::max(min_gram, size_t(1)),
      max_gram(std::max(max_gram, min_gram) {
  }

  size_t min_gram;
  size_t max_gram;
};